exonum/src/runtime/rust/api.rs

```diff
@@ -0,0 +1,155 @@
+    /// parameters. Message with the configuration parameters should be named as `Config`.
+    pub sources: Vec<ProtoSourceFile>,
+    /// List of service's proto include files.
+    pub includes: Vec<ProtoSourceFile>,
+}
+
+impl ArtifactProtobufSpec {
+    /// Creates a new artifact Protobuf specification instance from the given
+    /// list of Protobuf sources and includes.
+    pub fn new(
+        sources: impl IntoIterator<Item = impl Into<ProtoSourceFile>>,
+        includes: impl IntoIterator<Item = impl Into<ProtoSourceFile>>,
+    ) -> Self {
+        Self {
+            sources: sources.into_iter().map(|x| x.into()).collect(),
+            includes: includes.into_iter().map(|x| x.into()).collect(),
+        }
+    }
+}
+
+impl From<&(&str, &str)> for ProtoSourceFile {
+    fn from(v: &(&str, &str)) -> Self {
+        Self {
+            name: v.0.to_owned(),
+            content: v.1.to_owned(),
+        }
+    }
+}
+
+fn exonum_proto_sources() -> Vec<ProtoSourceFile> {
+    let proto = EXONUM_PROTO_SOURCES
+        .as_ref()
+        .iter()
+        .map(From::from)
+        .collect::<Vec<_>>();
+    let includes = EXONUM_INCLUDES
+        .as_ref()
+        .iter()
+        .map(From::from)
+        .collect::<Vec<_>>();
+
+    proto.into_iter().chain(includes).collect()
+}
+
+fn filter_exonum_proto_sources(
+    files: Vec<ProtoSourceFile>,
+    exonum_sources: Vec<ProtoSourceFile>,
+) -> Vec<ProtoSourceFile> {
+    files
+        .iter()
+        .cloned()
+        .filter(|file| !exonum_sources.contains(file))
+        .collect()
+}
+
+/// Returns API builder instance with the appropriate endpoints for the specified
+/// Rust runtime instance.
+pub fn endpoints(runtime: &RustRuntime) -> impl IntoIterator<Item = (String, ApiBuilder)> {
+    let artifact_proto_sources = runtime
+        .available_artifacts
+        .iter()
+        .map(|(artifact_id, service_factory)| {
+            (
+                artifact_id.clone(),
+                service_factory.artifact_protobuf_spec(),
+            )
+        })
+        .collect::<HashMap<_, _>>();
+    let exonum_sources = exonum_proto_sources();
+    // Cache filtered sources to avoid expensive operations in the endpoint handler.
+    let filtered_sources = artifact_proto_sources
+        .clone()
+        .into_iter()
+        .map(|(artifact_id, sources)| {
+            let mut proto = sources.sources;
+            proto.extend(filter_exonum_proto_sources(
+                sources.includes,
+                exonum_sources.clone(),
+            ));
+            (artifact_id, proto)
+        })
+        .collect::<HashMap<_, _>>();
+
+    let mut builder = ApiBuilder::new();
+    builder
+        .public_scope()
+        // This endpoint returns list of protobuf source files of the specified artifact,
+        // otherwise it returns source files of Exonum itself.
+        .endpoint("proto-sources", {
+            move |query: ProtoSourcesQuery| {
```

Comment:
Unnecessary but it would be good to define a return type of closure, I suppose

```suggestion
            move |query: ProtoSourcesQuery| -> Result<Vec<ProtoSourceFile>, api::Error> {
```